runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -5253,15 +5258,29 @@ class INTERPRETER_CLASS
 		_currentThread->callOutCount += 1;
 #endif /* JAVA_SPEC_VERSION >= 19 */
 		updateVMStruct(REGISTER_ARGS);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isInTrivialDownCall) {
+			VM_VMAccess::inlineExitVMToJNI(_currentThread);
+		}
+#else /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMAccess::inlineExitVMToJNI(_currentThread);
+#endif /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMHelpers::beforeJNICall(_currentThread);
 #if FFI_NATIVE_RAW_API
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values, values_raw);
 #else /* FFI_NATIVE_RAW_API */
 		ffiCallWithSetJmpForUpcall(_currentThread, cif, function, returnStorage, values);
 #endif /* FFI_NATIVE_RAW_API */
 		VM_VMHelpers::afterJNICall(_currentThread);
+#if JAVA_SPEC_VERSION >= 21
+		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
+		if (!_currentThread->isInTrivialDownCall) {
+			VM_VMAccess::inlineEnterVMFromJNI(_currentThread);
+		}
+#else /* JAVA_SPEC_VERSION >= 21 */
 		VM_VMAccess::inlineEnterVMFromJNI(_currentThread);
+#endif /* JAVA_SPEC_VERSION >= 21 */
```

Comment:
Somewhere after the downcall returns, `isInTrivialDownCall` must be reset to the value it had before this call, else the following pattern would be inappropriately disallowed:
```
	nonTrivialDowncall(); // invoked without LinkerOptions.IsTrivial.INSTANCE
		upcall1();
			trivialDowncall(); // invoked with LinkerOptions.IsTrivial.INSTANCE
		upcall2();
			legalDowncall(); // should not be disallowed
```

Reply:
This is not true as the `IsTrivial` option is only used by downcall and only set in applications to a given downcall `MethodHandle`. e.g.
```
MethodHandle mh = linker.downcallHandle(functionSymbol, fd, Linker.Option.isTrivial()); <-----
...
mh.invoke(...)
```
which eventually goes to our code at `openj9/internal/foreign/abi/InternalDowncallHandler.java`:
```
returnVal = invokeNative(linkerOpts.isTrivial(), <---------- check isTrivial() to set the value each time when performing downcall
getValidDowncallMemAddr(stateSegmt), retMemAddr, getValidDowncallMemAddr(downcallAddr), cifNativeThunkAddr, args);
```
That being said, `linkerOpts.isTrivial()` (for the same downcall MH) is checked every time when invoking the native so as to determine whether it is a trivial downcall or not, in which case any legal downcall/upcall will be allowed as long as the given downcall MH comes without `isTrivial` specified in applications.

Reply:
There is a problem, but my example was flawed. It isn't the second down-call that's the problem, it's the second up-call that will be disallowed, but should be allowed:
```
	nonTrivialDowncall(); // invoked without LinkerOptions.IsTrivial.INSTANCE
		upcall1();
			trivialDowncall(); // invoked with LinkerOptions.IsTrivial.INSTANCE
			// _currentThread->isInTrivialDownCall = TRUE;
		upcall2(); // should not be disallowed
```

Reply:
As discussed with Keith offline, we can reset `isInTrivialDownCall` to false after returning from the trival downcall in `inlInternalDowncallHandlerInvokeNative` given currentThread->isInTrivialDownCall is shared in such case.
```
		/* Remove VMaccess in the trivial downcall to benefit the JIT for better performance. */
		if (!_currentThread->isInTrivialDownCall) {
			VM_VMAccess::inlineEnterVMFromJNI(_currentThread);
			_currentThread->isInTrivialDownCall = FALSE; <---------
		}
```
Now the question whether a JNI upcall can be triggered & prevent from a trivial downcall (already brought the questions to OpenJDK via the mailing-list to see how they respond).

Reply:
Here're the response from OpenJDK as to the question above:
[1] Maurizio (The FFI Architect from Oracle) at https://mail.openjdk.org/pipermail/panama-dev/2023-July/019509.html
```
The behavior is undefined. It might crash, or not. <---------

While the description says that the downcall should not upcall, please 
note that the Javadoc doesn't say anywhere that ill-behaved downcalls 
will produce new errors. Moreover, the javadoc says this:

> Using this linker option when linking non trivial functions is likely 
> to have adverse effects, such as loss of performance, or JVM crashes.
>
Basically, you are in undefined behavior territory. It seems to me that 
you are inferring too much from the javadoc. <-------------

While I agree it might be desirable to detect this and log some kind of 
error (see email I sent yesterday), I don't think that, as thing stands, 
the javadoc is incorrect, and/or lacking information.
```
plus [2] Jorn (The FFI contributor from Oracle) at https://mail.openjdk.org/pipermail/panama-dev/2023-July/019513.html
```
So, you would only see a crash for the JNI upcall case only when using a 
debug build, but would see a crash always for the Panama upcall case.
```
against the latest announcement at https://mail.openjdk.org/pipermail/panama-dev/2023-July/019510.html at
```
There are several enhancements (aside from "regular" performance fixes) that we have been discussing 
over the years which can be delivered as follow-up JEPs:
* Integration between Linker and JNI [13] <--------
```
and FFI/JNI related discussion at [13]https://mail.openjdk.org/pipermail/panama-dev/2023-May/019079.html

-------------------------

Based on the response/announcement above, my conclusion is:
[1] Triggering JNI upcall from FFI downcall is totally undefined for the moment (which is out of FFI scope), in which case, for a trivial downcall,  Hotspot crashes the VM only with a **debug** build but lets it pass through to trigger a JNI upcall with a normal build (confirmed with my simple test at https://mail.openjdk.org/pipermail/panama-dev/2023-July/019508.html)
[2]Interaction of FFI with JNI will be discussed further and everything related to the topic will eventually go to a separate JEP to cover all cases with JNI.

Given such behavior is still undefined from FFI perspective and will be addressed in a follow-up JEP with JNI, we'd better avoid any unnecessary changes in the JNI related code (even though we are able to do so in theory) till the published JEP guides us how to deal with these cases with JNI.

Reply:
Just added the proposed code at https://github.com/eclipse-openj9/openj9/pull/17847#discussion_r1275510548
to reset the value of  `_currentThread->isInTrivialDownCall` after performing the downcall.

Reply:
I created a small test program (see https://github.com/keithc-ca/DowncallAbuse) that makes upcalls (via JNI) from a trivial downcall. Neither OpenJ9 (without this change) nor the RI had any objections nor any apparent problems running the code.

Reply:
Then we should keep it unchanged in such case to remain consistent with the RI behavior.